internal/store/builder.go

```diff
@@ -146,11 +154,37 @@ func (b *Builder) WithGenerateStoresFunc(f ksmtypes.BuildStoresFunc, u bool) {
 	b.useAPIServerCache = u
 }
 
+// WithGenerateCustomResourceStoresFunc configures a custom generate custom resource store function
+func (b *Builder) WithGenerateCustomResourceStoresFunc(f ksmtypes.BuildCustomResourceStoresFunc, u bool) {
+	b.buildCustomResourceStoresFunc = f
+	b.useAPIServerCache = u
```

Comment:
`b.useAPIServerCache` is used by `(store.*Builder).WithGenerateStoresFunc` as well, what would happen in the following scenario?

```go
store := ...
store.WithGenerateStoresFunc(..., false)
store.WithGenerateCustomResourceStoresFunc(..., true)
```

`(store.*Builder).useAPIServerCache` is implicitly overwritten by the invocation of `(store.*Builder).useAPIServerCache` though the expectation could be that they both keep a separate property / variable for this (e.g. `(store.*Builder).useAPIServerCacheForCustomResources`).

It could also be that `(store.*Builder).useAPIServerCache` should not be too tightly coupled with the function and thus needs to be configured separately by adding an extra method to do so.

Or this might be an indication the custom resources like this deserve their own builder, this way you wouldn't need to add respective methos for custom resources to the existing builder but create two separate builders and combine the stores later on.

Reply:
Thanks @Serializator for reminding! I have not even noticed that. I prefer this solution.
```
It could also be that (store.*Builder).useAPIServerCache should not be too tightly coupled with the function and thus needs to be configured separately by adding an extra method to do so.
```

Reply:
I agree when looking specifically at `(store.*Builder).useAPIServerCache` but I'd like to "widen" the "scope" of my initial feedback and look at the bigger picture in terms of extensibility of `store.Builder` itself.

I haven't looked through the issue or pull request completely so please correct me if I'm wrong! CR metrics are basically metrics like the built-in metrics by KSM but registrered a different way.

Why did you make the choice to extend the existing `store.Builder` rather than create a separate instance of it specifically for registering and working with CR metrics, and later on make some kind of "composite" store which combines the built-in and the CR metrics?

I'm asking in terms of future extensability. If there is ever a need to add an additional way of registering metrics then methods will again be added to the `store.Builder` to allow for it ("extending" the `store.Builder`) rather than "building upon it" and keepign the implementation of `store.Builder` light and clean.

I'd love to hear your point of view and opinion on this! I might be completely wrong 😉

Reply:
Hi @Serializator, it's really a good question.

I choose to extend `store.Builder` because IMO actually the Custom Resource is very similar to Kubernetes built-in resource. The only thing different is that we don't know their type, what their metrics look like, how to list and watch them(the client code is generated by `code-generator`). Once people tell us the necessary information, it behaves like a common built-in resource. It could be built and stored similarly, shared the same logic with `store.Builder`.

```
Rather than create a separate instance of it specifically for registering and working with CR metrics, and later on make some kind of "composite" store which combines the built-in and the CR metrics
```

I think this solution is more complicated. As I mentioned, the CR and built-in resource could share the same building logic once we have the information in `registry_factory.go`. Actually, what I'm thinking is all metrics could be registered by this interface including built-in resource.  From the aspect of extensability, the inner resource could be added by this method. Actually, I suppose removing the built-in resource registry from `store.Builder` and keep the CustomResourceStores may be more light and clean? 

Just some personal idea, lol 😄

Reply:
In that sense I understand and agree (not that you need my approval 😉) in terms of the extending the `store.Builder` and the why behind it.

`(store.Builder).buildCustomResourceStores` though is really similar to `(store.Builder).buildStores`, with the only difference being the "client" which is passed to the `listWatchFunc`. The crutch in this case is in the fact that `(store.Builder).buildStores` is aware of the client (`k8s.io/client-go/kubernetes.Interface` because `(store.Builder).buildStores` is responsible for calling the `listWatchFunc` function. Disappointing (😞 😆), maybe refactor in the future to abstract away the client (as this can differ now with the CR metrics).

Reply:
Hi @Serializator. You are right! We have the same opinions😃. That also distresses me! 

These two functions are so similar but unfortunately we have no ideas about what type the `customResourceClient` is. If we merge them together, all types of `listWatchFunc` in `internal/store` directory must be revised. This revision may be a little bit huge and does not focus on custom resource extension. So I decided to make a optimization in the future. That would be better for reviewing.

Reply:
I have added a `TODO` comment. We could optimize it later😉.